bin/propolis-server/src/lib/stats/virtual_machine.rs

```diff
@@ -0,0 +1,526 @@
+    kstat: &Kstat<'_>,
+    data: &Data<'_>,
+    instance_id: &str,
+) -> Option<i32> {
+    // Filter out anything that's not a `vmm:<instance>:vm` named kstat.
+    if kstat.ks_module != VMM_KSTAT_MODULE_NAME {
+        return None;
+    }
+    if kstat.ks_name != VM_KSTAT_NAME {
+        return None;
+    }
+    let Data::Named(named) = data else {
+        return None;
+    };
+
+    // Return the instance if the `vm_name` kstat matches our instance UUID.
+    if named.iter().any(|nd| {
+        if nd.name != VM_NAME_KSTAT {
+            return false;
+        }
+        let NamedData::String(name) = &nd.value else {
+            return false;
+        };
+        instance_id == *name
+    }) {
+        return Some(kstat.ks_instance);
+    }
+    None
+}
+
+// Produce `Sample`s for the `VcpuUsage` metric from the relevant kstats.
+fn produce_vcpu_usage<'a>(
+    vm: &'a VirtualMachine,
+    vcpu_stats: impl Iterator<Item = &'a (DateTime<Utc>, Kstat<'a>, Data<'a>)> + 'a,
+) -> Result<Vec<Sample>, Error> {
+    let mut out =
+        Vec::with_capacity(vm.n_vcpus as usize * N_VCPU_MICROSTATES as usize);
+    for (creation_time, kstat, data) in vcpu_stats {
+        let Data::Named(named) = data else {
+            return Err(Error::ExpectedNamedKstat);
+        };
+        let snapshot_time = hrtime_to_utc(kstat.ks_snaptime)?;
+
+        // Find the vCPU ID, from the `vmm:<instance>:vcpuX:vcpu` named kstat.
+        let vcpu_id = named
+            .iter()
+            .find_map(|named| {
+                if named.name == VCPU_KSTAT_PREFIX {
+                    named.value.as_u32().ok()
+                } else {
+                    None
+                }
+            })
+            .ok_or(Error::NoSuchKstat)?;
+
+        // We track each vCPU microstate starting with `time_`, and map them
+        // into our own definitions of the vCPU states. We need to aggregate all
+        // the occupancy times from the microstates that map to the same public
+        // state.
+        let mut occupancy_by_state = BTreeMap::new();
+        for nv in named.iter() {
+            // Skip kstats that are not known microstate names.
+            let Some(state) = kstat_microstate_to_state_name(nv.name) else {
+                continue;
+            };
+
+            // Get the current summed state occupancy, or insert one with 0.
+            let datum =
+                occupancy_by_state.entry(state.to_string()).or_insert_with(
+                    || Cumulative::with_start_time(*creation_time, 0),
+                );
+            *datum += nv.value.as_u64()?;
+        }
+
+        // Now convert the aggregated occupancy times into samples.
+        for (state, datum) in occupancy_by_state.into_iter() {
+            let metric = VcpuUsage { vcpu_id, state, datum };
+            let sample =
+                Sample::new_with_timestamp(snapshot_time, vm, &metric)?;
+            out.push(sample);
+        }
+    }
+    Ok(out)
```

Comment:
Would it be worthwhile to assert here that `out.len() <= original_capacity`? (I'm trying to think of good ways to produce a warning if we start collecting more statistics but don't bump the value of `N_VCPU_MICROSTATES`/whatever constant ultimately governs the size of the sample buffer.)

Reply:
> Hmm, I'm not sure what we'd be guarding against in that case. I had thought of this only as a value to pre-allocate with, but it seems like you're concerned about a correctness issue. Could you explain that more?

I don't think there's really a correctness issue as such, and in fact the question I meant to ask seems pretty fringy now that I think about it some more.

My real question here is about `KSTAT_LIMIT_PER_VCPU`: that constant reads to me as very lightly load-bearing in the sense that (per its doc comment) it's chosen so that the `KstatSampler` will buffer at least 64 samples even if the producer isn't being polled regularly. If we were to start querying for more states/collecting additional `Sample`s here that buffer would shrink. This is definitely not the end of the world, but it seems like the sort of thing that would be nice to catch if it happens, and this seemed like it might be a place where we could do that.

This is all definitely well into "optional/nice to have" territory, though--the code should work fine as-is.

Reply:
Ah I see. So one place we'd catch that is the self-statistics that the `KstatSampler` generates. It notes any dropped samples, and produces counters for them which themselves go into the timeseries database. The tradeoff there is really just space. We should have _some_ limit, but 64 is really arbitrary.

Reply:
Nice--if we already have some other drop detection then I think this is even less of a concern.